### DISCUSSION TITLE
Reducing Outliers - Chaining Strategies

### DISCUSSION THREAD

**Discussion by trivikram18** [MaartenGr/BERTopic#1866] [Q&A] [answered]
Hello, Maarten/ Hello All,

I have been trying to chain strategies as shown in the documentation below. 

![image](https://github.com/MaartenGr/BERTopic/assets/10544722/c0daadff-362b-4c3c-9179-043b17359dd0)

Is there any error in the documentation here, I see that in the first line, 'new_topics' are generated, and then in the subsequent line, 'topics' is passed as the parameter to the reduce_outliers. I thought these 2 lines were swapped. We would have created 'topics', and pass that to 'c-tf-idf' which returns 'new_topics', which is then passed to a 2nd strategy that returns the final topics.

The following are the relevant lines from my program ...

reduced_outliers_topics2 = topic_model.reduce_outliers(docs, original_topics, strategy="c-tf-idf", threshold=0.7)
reduced_outliers_topics = topic_model.reduce_outliers(docs, reduced_outliers_topics2, strategy="embeddings", threshold=0.7)

topic_model.update_topics(docs,
                          topics=reduced_outliers_topics,
                          representation_model=representation_model)

What I see is that it always takes the first strategy and seems to ignore the 2nd strategy. I tried running the program without the threshold value as well.

I am using BERTopic v0.16.0
Any help would be appreciated.

Thanks
Trivikram

**Comment by MaartenGr** [comment]
The documentation is indeed incorrect, it should be the following:

```python
# Use the "c-TF-IDF" strategy with a threshold
new_topics = topic_model.reduce_outliers(docs, new_topics , strategy="c-tf-idf", threshold=0.1)

# Reduce all outliers that are left with the "distributions" strategy
new_topics1 = topic_model.reduce_outliers(docs, new_topics , strategy="distributions")
```

That way, the reduced topics from the previous step should be used in the new step. If that does not work for you, it change the `threshold` value as it works differently depending on the underlying distribution of similarities.

**Reply by trivikram18** [reply]
Thanks, Maarten. I will try with various threshold values. 
I have a quick question though, does the chaining work only with 'c-tf-idf' and distributions only or can I choose any combination of the available 4 strategies?
When I ran the program with the few lines that I copied earlier I didn't see the desired result.

Regards
Trivikram

**Reply by MaartenGr** [reply, marked as answer]
You can use any combination of the available strategies however many times you want. If a strategy does not make any changes, then the threshold is set too high.

**Comment by trivikram18** [comment]
Hi Maarten,

One more question ... after the topic generation I see that the positive and negative topics/ words/ phrases are clustered into one. 
As an example, I see that 'I enjoyed the ride' and 'I did not enjoy the ride' both are clustered into one topic. Is there any way to separate these. Any technique or hyper-parameters that I need to tune to get them right?

Thanks again.

**Reply by MaartenGr** [reply]
Generally, these embeddings focus on their semantic representation and to a lesser extent the underlying sentiment. It is not that they do not capture it but merely that is a lesser focus of attention. Instead, it might be worthwhile to use something like [semi-supervised topic modeling](https://maartengr.github.io/BERTopic/getting_started/semisupervised/semisupervised.html) in BERTopic or even extend the embeddings to capture and focus more on the sentiment. Another trick is to simply separate the topics with a sentiment-based algorithm afterwards and then run back through [manual](https://maartengr.github.io/BERTopic/getting_started/manual/manual.html) BERTopic.

**Reply by trivikram18** [reply]
Thanks for those suggestions Maarten. I am keen on the last suggestion of using manual BERTopic, however I am lacking a little detail on it. 
1. Run the BERTopic on the data and generate topics
2. Separate the docs or topics. I wasn't sure how to proceed after this. 

One more approach I was thinking of is to separate the docs based on the sentiment, run 2 models one for each of positive and negative, and then merge the models.

Thanks

**Reply by MaartenGr** [reply]
If you separate the documents, you will have essentially created labeled data. You can use that labeled data together with manual BERTopic to create a model that differentiates between positive and negative sentiment. 

To make this concrete. If you have 3 topics generated and you split them based on their sentiment (positive vs. negative), you will now have 6 topics (1-positive, 1-negative, 2-positive, 2-negative, 3-positive, 3-negative). These 6 topics can serve as labels in manual BERTopic. 

> One more approach I was thinking of is to separate the docs based on the sentiment, run 2 models one for each of positive and negative, and then merge the models.

That might also work!

**Reply by trivikram18** [reply]
Thanks a lot, Maarten, When I merge the models and get the get_topic_info, I see that one topic is missing. From the following screenshot, we are missing Topic 10. I am investigating this and so far I didn't identify the issue. This is the first time I am using BERTopic.merge_models(). 

![image](https://github.com/MaartenGr/BERTopic/assets/10544722/453c7e48-a7cd-423e-9db0-90ecf20d79e8)

**Comment by trivikram18** [comment]
Hi Maarten, any insight on this issue of missing topic? I couldn't find the issue in my code.

**Reply by MaartenGr** [reply]
Could you share your **full** code? That makes it easier to detect what is happening. Also, did you try installing BERTopic from the main branch? I believe there might be a fix for this there.

**Reply by trivikram18** [reply]
Hello Maarten,

I have not tried the version from the main branch yet. 

The following is the code (some parts like reading the data file etc is eliminated for simplicity)

`
# Function to build the model instance with various parameters

def get_topic_model(embedding_model):

    # Initiate UMAP
    umap_model = UMAP(n_neighbors = params['n_neib'],
                      n_components = params['n_comp'],
                      min_dist = params['min_dist'],
                      metric = params['umap_metric'],
                      random_state = 142)

    # HDBSCAN
    hdbscan_model = HDBSCAN(min_cluster_size=params['min_cluster_size'],
                            min_samples=params['min_samples'], prediction_data=True)

    vectorizer_model = CountVectorizer(stop_words=None, ngram_range=params['n_gram'])

    ctfidf_model = ClassTfidfTransformer(bm25_weighting=True,
                                         reduce_frequent_words=True)

    # Create your representation model
    representation_model = MaximalMarginalRelevance(diversity=params['diversity'])

    topic_model = BERTopic(# language="multilingual",
                            calculate_probabilities = True,
                            min_topic_size = params['min_topic_size'],
                            top_n_words = params['top_n_words'],
                            n_gram_range = params['n_gram'],
                            verbose = True,
                            nr_topics=params['nr_topics'],
                            # Step 1 - Extract embeddings
                            embedding_model = embedding_model,
                            # Step 2 - Reduce dimentionality
                            umap_model = umap_model,
                            # Step 3 - Cluster reduced embeddings
                            hdbscan_model = hdbscan_model,
                            # Step 4 - Tokenize topics
                            vectorizer_model = vectorizer_model,
                            # Step 5 - Extract topics
                            ctfidf_model = ctfidf_model,
                            # Step 6 - (optional) Fine-tune topic representations
                            representation_model = representation_model
                            )
    return topic_model


def generate_topics(data_frame, id_var, verb_var, params, sentiment):

    data_frame = data_frame.loc[data_frame['sentiment_labels'].isin(sentiment)]

    data_frame.reset_index(drop=True, inplace=True)

	docs = data_frame['verb_var']

	print (f"# of verbs {len(docs)}")

	min_topic_size = int(len(docs) * 0.005)

	if min_topic_size < 10:
		min_topic_size = 10
	elif min_topic_size > 60:
		min_topic_size = 60

	# To solve ValueError: k must be less than or equal to the number of training points
	if len(docs) < min_topic_size:
		min_topic_size = len(docs)

	print (f"topic_size {min_topic_size}")
	params['min_topic_size'] = min_topic_size

	print ("Load the embedding model files")
	embedding_model = SentenceTransformer("all-MiniLM-L6-v2")

	print ("Create model instance")
	topic_model = get_topic_model(embedding_model, params)

	representation_model = topic_model.representation_model

	# Run BERTopic model
	print ("Fit the model:")
	original_topics, probabilities = topic_model.fit_transform(docs)

	# Reduce outliers
	if -1 in original_topics:
		print ("Reducing outliers:")
		try:
			# Chaining strategies
			reduced_outliers_topics2 = topic_model.reduce_outliers(docs, original_topics, strategy="embeddings", threshold=0.6)
			reduced_outliers_topics = topic_model.reduce_outliers(docs, reduced_outliers_topics2, strategy="c-tf-idf")

			topic_model.update_topics(docs,
									  topics=reduced_outliers_topics,
									  representation_model=representation_model)

		except ValueError:
			reduced_outliers_topics = original_topics
	else:
		print ("No outliers to reduce")
		reduced_outliers_topics = original_topics

	return topic_model, reduced_outliers_topics

params = {"n_neib": 15, "n_comp": 10, "min_dist": 0.0,
		  "umap_metric": 'cosine', "min_cluster_size": 20,
		  "min_samples": 10, "n_gram": (2,3), "top_n_words": 10,
		  'diversity': 0.7, 'nr_topics': "auto"}

verb_var = "q3"

q3_positive, q3_p_topics = generate_topics(df, id_var, verb_var, params, ['Positive',])
q3_p_tinfo = q3_positive.get_topic_info()

q3_non_positive, q3_nonp_topics = generate_topics(df, id_var, verb_var, params, ['Negative', 'Neutral'])
q3_nonp_tinfo = q3_non_positive.get_topic_info()

q3_topic_model = BERTopic.merge_models([q3_positive, q3_non_positive], min_similarity=0.7)
q3_tinfo = q3_topic_model.get_topic_info()

docs = df['verb_var']

topics, probs = q3_topic_model.transform(docs)
print (np.unique(topics))
print (len(np.unique(topics)))
q3_tinfo = q3_topic_model.get_topic_info()
print (q3_tinfo['Topic'].values)
print (len(q3_tinfo))
`

**Reply by MaartenGr** [reply]
It's a bit difficult to read due to the formatting you used but I cannot find something that looks out of place. Could you try using BERTopic from its main branch?

**Reply by trivikram18** [reply]
Hi Maarten, Sorry for the formatting. I did try fixing the formatting before posting but somehow couldn't get it right. Thanks for reviewing it anyway. Sure, I will install it from the main branch.

**Reply by trivikram18** [reply]
Hello Maarten, 

I just wanted to drop an update on the missing topic from the merged model. Installing from the main branch doesn't solve the issue. I installed it in a new environment and tested it.
Upon further investigation, it seems like the issue occurs when we reduce the outliers of the individual models before merging them. If I don't reduce the outliers then everything seems fine. There is no missing topic.

Thanks

**Reply by MaartenGr** [reply]
Thanks for the update! I hoped installing from the main branch would resolve this issue. Do you have an example that is completely reproducible? That way, I could take some time myself to figure out what causes this if I can reproduce it myself.

**Reply by trivikram18** [reply]
When I try with another dataset I see the issue again (the missing topic) even when I don't reduce the outliers.

**Reply by MaartenGr** [reply]
Thanks for sharing. Is that a dataset you can share? A reproducible example would help me find the issue more easily and resolve it.

**Reply by trivikram18** [reply]
The attached archive has the reproducible code and data. Very much appreciate your help. 

[topic_modelling.zip](https://github.com/MaartenGr/BERTopic/files/14804711/topic_modelling.zip)

**Reply by MaartenGr** [reply]
@trivikram18 Thank you for the reproducible code and data! It helped me understand what the issue might be and how to fix it. It took a while to figure out but it seemed that I made a silly mistake in the code (of which there are plenty 😅) but I think I managed to fix it with https://github.com/MaartenGr/BERTopic/pull/1898 

Could you try that out and see if it works? Based on my experiments it does but I want to be sure.

**Reply by trivikram18** [reply]
Hi Maarten, 

Thanks for checking. Silly mistakes; we all do them...

I still see something strange happening here. Now, the number of topics is the same both from the data and the topic_info however the numbering of the topics doesn't seem to be a match. 

From the following image we can see that there are 17 topics (ranging from -1 to 15)
![image](https://github.com/MaartenGr/BERTopic/assets/10544722/f8953777-b135-4271-8824-494dff708e98)

However, from the following image the topics (coming from topic_info) are still 17 in number but topic number 3 is missing and topic number 16 is additional.
![image](https://github.com/MaartenGr/BERTopic/assets/10544722/8fc7f3db-03f9-418a-973f-a6c9e16b53ef)

**Reply by MaartenGr** [reply]
Hmmm, that's strange. I see different behavior. Are you running the **exact** same code as you did before?

**Reply by trivikram18** [reply]
Yes. exactly the same program. I created a new environment, installed BERTopic from the following link, and executed the program.

https://github.com/MaartenGr/BERTopic.git@fix_1866

**Reply by MaartenGr** [reply]
@trivikram18 Sorry for the late reply, it seems I made a silly mistake in that PR which should now be fixed. Can I ask you (hopefully the last time) to install from the PR and check whether it works?

**Reply by trivikram18** [reply]
Hi Maarten,

Thanks for looking into it. I tested this on 2 datasets and it looks good now. 
Is there any way we can get the representative docs for the merged model?
Also, are we getting a new version of BERTopic on Pypi anytime soon?

Best wishes.

**Reply by MaartenGr** [reply]
> Is there any way we can get the representative docs for the merged model?

That is currently not possible as I intended the `.merge_models` to be somewhat feasible for federated learning, which means training data should not be passed around. It potentially could be added in a later version but it is currently not on my priority list. 

> Also, are we getting a new version of BERTopic on Pypi anytime soon?

I intend to create a new minor release as soon as I know that [this PR](https://github.com/MaartenGr/BERTopic/pull/1898) works alright. It is the PR that I created to solve your problem which I updated a couple of days ago. As soon as I know it works for you and potentially @vincemu who has checked out the PR before, I can merge it and start on the official release.

**Reply by trivikram18** [reply]
Thanks for the clarification Maarten. This PR has solved the issue.
